File: src/main/java/ru/hollowhorizon/hc/client/render/game/GPUMemoryManager.java
package ru.hollowhorizon.hc.client.render.game;

import net.minecraft.client.Minecraft;
import org.lwjgl.opengl.GL15;
import org.lwjgl.opengl.GL30;
import ru.hollowhorizon.hc.HollowCore;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.stream.Collectors;

public class GPUMemoryManager {
    public static GPUMemoryManager INSTANCE = new GPUMemoryManager();

    private final ConcurrentMap<WeakReference<VBO>, Integer> VBOs = new ConcurrentHashMap();
    private final ConcurrentMap<WeakReference<VAO>, Integer> VAOs = new ConcurrentHashMap();

    private GPUMemoryManager()
    {
    }

    public void initialize()
    {
        Minecraft.getInstance().execute(new ClearingRunnable(this));
    }

    public VAO createVAO() {
        final int id = GL30.glGenVertexArrays();
        final VAO vao = new VAO(id);

        VAOs.put(new WeakReference<>(vao), id);

        return vao;
    }

    public VBO createVBO(final int type){
        final int id = GL15.glGenBuffers();
        final VBO vbo = new VBO(id, type);

        VBOs.put(new WeakReference<>(vbo), id);

        return vbo;
    }

    private static final class ClearingRunnable implements Runnable {

        private final GPUMemoryManager managerToHandle;

        public ClearingRunnable(final GPUMemoryManager managerToHandle) {this.managerToHandle = managerToHandle;}

        @Override
        public void run() {
            final List<Map.Entry<WeakReference<VBO>, Integer>> removedVBOs =
                    managerToHandle.VBOs.entrySet().stream().filter(entry -> entry.getKey().get() == null).collect(Collectors.toList());
            final List<Map.Entry<WeakReference<VAO>, Integer>> removedVAOs =
                    managerToHandle.VAOs.entrySet().stream().filter(entry -> entry.getKey().get() == null).collect(Collectors.toList());

            removedVBOs.forEach(weakReferenceIntegerEntry -> {
                managerToHandle.VBOs.remove(weakReferenceIntegerEntry.getKey());
                GL15.glDeleteBuffers(weakReferenceIntegerEntry.getValue());
            });

            removedVAOs.forEach(weakReferenceIntegerEntry -> {
                managerToHandle.VAOs.remove(weakReferenceIntegerEntry.getKey());
                GL30.glDeleteVertexArrays(weakReferenceIntegerEntry.getValue());
            });

            //Reschedule clearing task.
            //We will have to wait a minimal amount of time.
            final Thread rescheduleThread = new Thread(() -> {
                try
                {
                    Thread.sleep(500);
                }
                catch (final InterruptedException ignored)
                {
                }

                try
                {
                    Minecraft.getInstance().execute(this);
                }
                catch (final Exception ex)
                {
                    HollowCore.LOGGER.error("Failed to reregister the GPUManager. GPU memory leaks will occur.", ex);
                }
            });
            rescheduleThread.start();
        }

        @Override
        public String toString()
        {
            return "ClearingRunnable{" +
                    "managerToHandle=" + managerToHandle +
                    '}';
        }
    }

    @Override
    public String toString()
    {
        return "GPUManager{" +
                "VBOs=" + VBOs +
                ", VAOs=" + VAOs +
                '}';
    }
}

File: src/main/java/ru/hollowhorizon/hc/client/graphics/GPUMemoryManager.kt
package ru.hollowhorizon.hc.client.graphics

import net.minecraft.client.Minecraft
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL30
import ru.hollowhorizon.hc.HollowCore
import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors


@OnlyIn(Dist.CLIENT)
class GPUMemoryManager private constructor() {
    private val vboMap = ConcurrentHashMap<WeakReference<VBO?>, Int>()
    private val vaoMap = ConcurrentHashMap<WeakReference<VAO?>, Int>()

    fun initialize() {
        Minecraft.getInstance().execute(ClearingRunnable(this))
    }

    fun createVAO(): VAO {
        val id = GL30.glGenVertexArrays()
        val vao = VAO(id)
        vaoMap[WeakReference(vao)] = id
        return vao
    }

    fun createVBO(type: Int): VBO {
        val id = GL15.glGenBuffers()
        val vbo = VBO(id, type)
        vboMap[WeakReference(vbo)] = id
        return vbo
    }

    private class ClearingRunnable(private val managerToHandle: GPUMemoryManager) : Runnable {
        override fun run() {
            val removedVBOs = managerToHandle.vboMap.entries.stream()
                .filter { it.key.get() == null }
                .collect(Collectors.toList())
            val removedVAOs = managerToHandle.vaoMap.entries.stream()
                .filter { it.key.get() == null }
                .collect(Collectors.toList())

            removedVBOs.forEach {
                managerToHandle.vboMap.remove(it.key)
                GL15.glDeleteBuffers(it.value)
            }
            removedVAOs.forEach {
                managerToHandle.vaoMap.remove(it.key)
                GL15.glDeleteBuffers(it.value)
            }

            val rescheduleThread = Thread {
                Thread.sleep(500)
                try {
                    Minecraft.getInstance().execute(this)
                } catch (ex: Exception) {
                    HollowCore.LOGGER.error("Failed to clear the GPUManager. GPU memory leaks will occur.", ex)
                }
            }
            rescheduleThread.start()
        }
    }

    companion object {
        val instance = GPUMemoryManager()
    }
}